carsus/io/cmfgen/base.py

```diff
@@ -0,0 +1,440 @@
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute.
+    """
+
+    keys = ['!Date',  # Metadata to parse from header. TODO: look for more keys
+            '!Format date',
+            '!Number of energy levels',
+            '!Ionization energy',
+            '!Screened nuclear charge',
+            '!Number of transitions',
+            ]
+
+    def load(self, fname):
+
+        meta = parse_header(fname, self.keys)
+        kwargs = {}
+        kwargs['header'] = None
+        kwargs['index_col'] = False
+        kwargs['sep'] = '\s+'
+        kwargs['skiprows'] = find_row(fname, "Number of transitions", num_row=True)
+
+        n = int(meta['Number of energy levels'])
+        kwargs['nrows'] = n
+
+        columns = ['Configuration', 'g', 'E(cm^-1)', 'eV', 'Hz 10^15', 'Lam(A)']
+
+        try:
+            df = pd.read_csv(fname, **kwargs, engine='python')
+
+        except pd.errors.EmptyDataError:
+            df = pd.DataFrame(columns=columns)
+            warnings.warn('Empty table')
+
+        # Assign column names by file content
+        if df.shape[1] == 10:
+            # Read column names and split them keeping one space (e.g. '10^15 Hz')
+            columns = find_row(fname, 'E(cm^-1)', "Lam").split('  ')
+            # Filter list elements containing empty strings
+            columns = [c for c in columns if c != '']
+            # Remove left spaces and newlines
+            columns = [c.rstrip().lstrip() for c in columns]
+            columns = ['Configuration'] + columns
+            df.columns = columns
+
+        elif df.shape[1] == 7:
+            df.columns = columns + ['#']
+            df = df.drop(columns=['#'])
+
+        elif df.shape[1] == 6:
+            df.columns = range(6)  # FIXME: These files don't have column names
+
+        elif df.shape[1] == 5:
+            df.columns = columns[:-2] + ['#']
+            df = df.drop(columns=['#'])
+
+        else:
+            warnings.warn('Inconsistent number of columns')  # TODO: raise exception here (discuss)
+
+        self.fname = fname
+        self.base = df
+        self.columns = df.columns.tolist()
+        self.meta = meta
+
+    def to_hdf(self, key='/energy_levels'):
+        with pd.HDFStore(self.fname + '.h5', 'a') as f:
+            f.append(key, self.base, format='table', data_columns=self.columns)
+            f.get_storer(key).attrs.metadata = self.meta
+
+
+class CMFGENOscillatorStrengthsParser(BaseParser):
+    """
+        Description
+        ----------
+        base : pandas.DataFrame
+        columns : list of str
+            (default value = ['State A', 'State B', 'f', 'A', 'Lam(A)', 'i', 'j', 'Lam(obs)', '% Acc'])
+        meta : dict
+            Metadata parsed from file header.
+
+        Methods
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute.
+    """
+
+    keys = CMFGENEnergyLevelsParser.keys
```

Comment:
just subclass from a `BaseCMFGen` parser

Reply:
But only *energy levels*  and *oscillator strengths* parser have the same keys, because both tables lives in the same OSC file.